marlowe-playground-client/src/Blockly/Dom.purs

```diff
@@ -0,0 +1,253 @@
+  -- are the parsed <next> elements which represent the siblings of a parent Statement.
+  readAsBlock :: forall m. MonadEffect m => MonadThrow ReadDomError m => Element -> m (Tuple Block (Array Block))
+  readAsBlock element =
+    if Element.tagName element /= "block" then
+      throwError $ TypeMismatch element "block"
+    else do
+      blockId <- liftEffect $ Element.id element
+      blockType <- getAttribute "type" element
+      elementChildren <- getChildrens element
+      -- Parse each child element, which can be Either a BlockChild (field, value, statement) or a list of
+      -- siblings to pass to my parent
+      { left, right } <- separate <$> traverse readAsBlockChild elementChildren
+      let
+        children = Object.fromFoldable left
+      pure
+        $ Tuple
+            (Block { id: blockId, type: blockType, children })
+            (join right)
+
+  -- Tries to read an Element as the children of a Block node. The Left hand side of the return represents a Block direct children
+  -- while the Right hand side represents the parsed `<next>` elements which should be appended to a parents Statement.
+  readAsBlockChild ::
+    forall m.
+    MonadEffect m =>
+    MonadThrow ReadDomError m =>
+    Element ->
+    m (Either ChildWithName (Array Block))
+  readAsBlockChild element = do
+    case Element.tagName element of
+      "field" -> do
+        name <- getAttribute "name" element
+        value <- getElementText element
+        pure $ Left $ Tuple name $ Field value
+      "statement" -> do
+        name <- getAttribute "name" element
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        pure $ Left $ Tuple name $ Statement ([ block ] <> nexts)
+      "value" -> do
+        name <- getAttribute "name" element
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        -- Even if a value can have a Block appended, it should have a single block child and no "siblings"
+        if length nexts /= 0 then
+          throwError $ IncorrectSiblingNesting "value"
+        else
+          pure $ Left $ Tuple name $ Value block
+      "next" -> do
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        pure $ Right ([ block ] <> nexts)
+      _ -> throwError $ TypeMismatch element "field, statement, value, next"
+
+  getSingleChild :: forall m. MonadEffect m => MonadThrow ReadDomError m => Element -> m Element
+  getSingleChild element = do
+    children <- getChildrens element
+    if length children /= 1 then
+      throwError $ SingleChildExpected element $ length children
+    else do
+      pure $ unsafePartial $ UnsafeArray.head children
+
+  getChildrens :: forall m. MonadEffect m => Element -> m (Array Element)
+  getChildrens element =
+    liftEffect do
+      elements <- (ParentNode.children <<< Element.toParentNode) element
```

Comment:
I think this would be equivalent to:
```haskell
elements <- ParentNode.children $ Element.toParentNode element
```
Or else:
```haskell
elements <- ParentNode.children (Element.toParentNode element)
```

Reply:
I personally prefer the use of `<<<` but I'm 💯 open to the others if that's what people prefer

Reply:
I don't have a problem with `<<<`, I just thought it was making things complicated here because it doesn't save any parenthesis. I didn't expect it to be controversial. If it makes it easier to read for you guys we can leave it as is

Reply:
yeah I just mean there are 3 ways to write this that are _almost_ the same and yet the three of us each prefer a different one :)